refactor VM- headers handling, use webRequest only when needed

tophf opened the pull request:
Attaches onBeforeSendHeaders listener only when a GM_xhr is active, automatically detaches when all done.

tophf commented on src/background/utils/requests.js:
```diff
@@ -188,58 +191,36 @@ function decodeBody(obj) {
+  verify[requestId] = reqId;
+  req.coreId = requestId;
+  // if the listener is to be unregistered, it must occur in the next event loop cycle
+  setTimeout(transformHeaders, 0, reqId);
```
Hmm, I wonder if onBeforeSendHeaders may be called multiple times during redirects? If so I need to unregister later, probably by using onResponseStarted and/or a bigger timeout.

tophf commented on src/background/utils/requests.js:
```diff
@@ -188,58 +193,37 @@ function decodeBody(obj) {
+  req.coreId = requestId;
+  return {
+    requestHeaders: headers
+    .filter(({ name }) => name !== VM_VERIFY && (!req.anonymous || !/^cookie$/i.test(name)))
```
I wonder if `filter` should come after `concat` so that a `cookie` is removed from custom headers as well? OTOH it's the userscript's responsibility to provide sensible noncontradictory params...

tophf replied:
Nevermind, previously they were filtered so I'll just do it here as well.

tophf commented:
Added Stage-3 optional?.chaining?.operator :wink: hopefully you don't mind.

gera2ld commented on .babelrc.js:
```diff
@@ -16,5 +16,8 @@ module.exports = {
     '@babel/plugin-proposal-function-bind',
+    // `loose:true` produces `foo==null` and it's fine because we'll never use `document.all`
+    // which is the only reason for the default `false`
+    ['@babel/plugin-proposal-optional-chaining', { loose: true }],
```
This is already supported in `@gera2ld/plaid/config/babelrc`.

tophf replied:
Great!

gera2ld commented on package.json:
```diff
@@ -18,10 +18,12 @@
   "description": "Violentmonkey",
   "devDependencies": {
     "@babel/plugin-proposal-function-bind": "^7.2.0",
+    "@babel/plugin-proposal-optional-chaining": "^7.6.0",
```
We don't need this.

gera2ld commented on package.json:
```diff
@@ -22,6 +22,7 @@
     "@gera2ld/plaid": "~1.4.12",
     "@gera2ld/plaid-vue": "~1.4.9",
     "@gera2ld/plaid-webpack": "~1.5.0",
+    "@types/chrome": "latest",
```
`yarn.lock` should be updated.

tophf replied:
Done.
P.S. turns out my IDE was using npm by default so I've switched it to yarn now.

tophf commented:
Had to force-push a few times because I accidentally added a bug yesterday while trying to harden the conditions for the listener to be registered. Fixed that and added a few comments that clarify the logic, hopefully.
